### DISCUSSION TITLE
Question about function evaluation

### DISCUSSION THREAD

**Discussion by njcstreet** [ClickHouse/ClickHouse#60307] [Q&A] [answered]
Hi,

We have a use case for ClickHouse that requires fairly complex nesting of conditional statements. Can anyone tell me, if I define an expression like item_count below which itself may do some fairly complicated work (simple example in this case of counting the number of elements in a string based on a separator), if I then re-use this field multiple times in an if statement, will that expression be re-evaluated every time it is encountered? Or will ClickHouse evaluate item_count once for each row and re-use the result each time it is then used in any subsequent expression? To put it another way, if I run the statement below, will length(splitByChar('|', my_field))  be evaluated once or twice for each row?

`select length(splitByChar('|', my_field)) as item_count
, if(item_count = 3, 'Three Items',	if(item_count > 3, 'More than three items', 'Less than three items')) as description
from my_table i `

Many thanks,

Nigel

**Comment by den-crane** [comment, marked as answer]
Expressions are reduced to minimal subexpressions and evaluated only one time. 
It does not matter if you use aliases or not.

`f1(f2(x))+f2(x)` -- f2 is evaluated only one time,

`length(splitByChar('|', my_field)) as item_count` all these will be evaluated only once, even if you write it this way:

```
select length(splitByChar('|', my_field)) as item_count , 
     if(   length(splitByChar('|', my_field))   = 3, 'Three Items',	
            if(length(splitByChar('|', my_field))  > 3, 'More than three items', 'Less than three items')) as description 
from my_table i 
```
BTW you can simplify it using `multiIf`.

**Reply by njcstreet** [reply]
Perfect, thanks very much for your response.
